sympy/matrices/expressions/matexpr.py

```diff
@@ -708,7 +715,10 @@ def __new__(cls, name, n, m):
                 return name[n, m]
         if isinstance(name, str):
             name = Symbol(name)
-        name = _sympify(name)
+        else:
+            name = _sympify(name)
+            if not isinstance(name.kind, MatrixKind):
+                raise TypeError("First argument of MatrixElement should be a matrix")
```

Comment:
This is what is needed to prevent bugs like this in future. When this exception is raised you can look back to the caller to see why it is attempting an invalid substitution